File: cses/leetcode/go/974.go
package main

import "fmt"

func subarraysDivByK(nums []int, k int) int {
	n := len(nums)
	pref := make([]int, n)
	mp := make(map[int]int)
	for i := 0; i < n; i++ {
		if i == 0 {
			pref[i] = nums[i]
		} else {
			pref[i] = pref[i-1] + nums[i]
		}
		rem := pref[i] % k
		mp[rem] += 1
	}
	//print the prefix sum
	fmt.Println(pref)
	fmt.Println(mp)
	sum := 0
	for key, val := range mp {
		if key == 0 {
			sum += val
			sum += (val * (val - 1)) / 2
		} else {
			sum += (val * (val - 1)) / 2
		}
	}
	fmt.Println("Sum = ", sum)
	return sum
}

File: cses/leetcode/cpp/974.cpp
#include <bits/stdc++.h>
using namespace std;
typedef long long ll;

class Solution
{
public:
    int subarraysDivByK(vector<int> &nums, int k)
    {
        int n = nums.size();
        map<int, int> mp;
        vector<int> pref(n);
        // construct the prefix sum
        for (int i = 0; i < n; i++)
        {
            if (i == 0)
            {
                pref[i] = nums[i];
            }
            else
            {
                pref[i] = pref[i - 1] + nums[i];
            }
            int rem = pref[i] % k;
            mp[rem]++;
        }
        int count = 0;
        for (auto it : mp)
        {
            count += ((it.second) * (it.second - 1)) / 2;
            if (it.first == 0 and it.second == 1)
            {
                count = 1;
            }
        }
        return count;
    }
};